File: document_scanner_flutter/ios/document_scanner_flutter/Sources/document_scanner_flutter/StabilityTracker.swift
import Foundation

struct StabilityUpdate {
  let state: String
  let progress: Double
  let stableFrames: Int64
  let shouldCapture: Bool
  let corners: [[String: NSNumber]]?

  init(
    state: String,
    progress: Double,
    stableFrames: Int64,
    shouldCapture: Bool,
    corners: [[String: NSNumber]]? = nil
  ) {
    self.state = state
    self.progress = progress
    self.stableFrames = stableFrames
    self.shouldCapture = shouldCapture
    self.corners = corners
  }
}

final class StabilityTracker {
  private var distanceThreshold = 50.0
  private var delayNanoseconds: UInt64 = 1_000_000_000
  private var durationNanoseconds: UInt64 = 1_000_000_000
  private var cooldownNanoseconds: UInt64 = 1_500_000_000

  private var previous: [Double]?
  private var displayed: [Double]?
  private var pendingMovement: [Double]?
  private var stableSince: UInt64 = 0
  private var stableFrames: Int64 = 0
  private var cooldownUntil: UInt64 = 0
  private var capturedCurrentDocument = false
  private var lastDetectedAt: UInt64 = 0

  // A single failed frame is common with a live camera. Keep the current
  // overlay briefly so the document border does not flash on and off.
  private let lostDetectionGraceNanoseconds: UInt64 = 250_000_000

  init(options: [String: Any]?) {
    updateOptions(options)
  }

  func updateOptions(_ options: [String: Any]?) {
    distanceThreshold = number(options, "autoCaptureDistanceThreshold", distanceThreshold)
    delayNanoseconds = milliseconds(
      number(options, "autoCaptureDelayMs", Double(delayNanoseconds) / 1_000_000),
      minimum: 0)
    durationNanoseconds = milliseconds(
      number(options, "autoCaptureDurationMs", Double(durationNanoseconds) / 1_000_000),
      minimum: 1)
    cooldownNanoseconds = milliseconds(
      number(options, "autoCaptureCooldownMs", Double(cooldownNanoseconds) / 1_000_000),
      minimum: 0)
  }

  func update(
    corners: [[String: NSNumber]]?,
    imageWidth: Int,
    imageHeight: Int,
    timestamp: UInt64,
    autoCaptureEnabled: Bool
  ) -> StabilityUpdate {
    guard let values = flattened(corners) else {
      if previous != nil,
        timestamp >= lastDetectedAt,
        timestamp - lastDetectedAt <= lostDetectionGraceNanoseconds
      {
        return progressUpdate(
          timestamp: timestamp,
          autoCaptureEnabled: autoCaptureEnabled,
          allowCapture: false)
      }
      let hadDocument = previous != nil
      resetCandidate()
      capturedCurrentDocument = false
      return StabilityUpdate(
        state: hadDocument ? "lost" : "searching",
        progress: 0,
        stableFrames: 0,
        shouldCapture: false)
    }

    lastDetectedAt = timestamp

    guard let prior = previous else {
      previous = values
      displayed = values
      stableSince = timestamp
      stableFrames = 1
      return StabilityUpdate(
        state: "detected",
        progress: 0,
        stableFrames: 1,
        shouldCapture: false,
        corners: mapped(displayed))
    }

    if maximumMovement(prior, values, imageWidth, imageHeight) >= distanceThreshold {
      if let pendingMovement,
        maximumMovement(pendingMovement, values, imageWidth, imageHeight) < distanceThreshold
      {
        // Two consecutive, mutually close frames represent real movement.
        // A single distant contour is detector noise and must not reset the
        // progress.
        previous = values
        displayed = smoothed(displayed, toward: values)
        self.pendingMovement = nil
        stableSince = timestamp
        stableFrames = 1
        capturedCurrentDocument = false
        return StabilityUpdate(
          state: "detected",
          progress: 0,
          stableFrames: 1,
          shouldCapture: false,
          corners: mapped(displayed))
      }
      pendingMovement = values
      return progressUpdate(
        timestamp: timestamp,
        autoCaptureEnabled: autoCaptureEnabled,
        allowCapture: false)
    }

    pendingMovement = nil
    previous = values
    displayed = smoothed(displayed, toward: values)
    stableFrames += 1
    return progressUpdate(
      timestamp: timestamp,
      autoCaptureEnabled: autoCaptureEnabled,
      allowCapture: true)
  }

  private func progressUpdate(
    timestamp: UInt64,
    autoCaptureEnabled: Bool,
    allowCapture: Bool
  ) -> StabilityUpdate {
    let elapsed = timestamp >= stableSince ? timestamp - stableSince : 0
    if elapsed < delayNanoseconds {
      return StabilityUpdate(
        state: "detected",
        progress: 0,
        stableFrames: stableFrames,
        shouldCapture: false,
        corners: mapped(displayed ?? previous))
    }

    let scanningElapsed = elapsed - delayNanoseconds
    let progress = min(1, Double(scanningElapsed) / Double(max(1, durationNanoseconds)))
    if scanningElapsed < durationNanoseconds {
      return StabilityUpdate(
        state: "stabilizing",
        progress: progress,
        stableFrames: stableFrames,
        shouldCapture: false,
        corners: mapped(displayed ?? previous))
    }

    let shouldCapture = allowCapture
      && autoCaptureEnabled
      && !capturedCurrentDocument
      && timestamp >= cooldownUntil
    if shouldCapture { capturedCurrentDocument = true }
    return StabilityUpdate(
      state: "stable",
      progress: 1,
      stableFrames: stableFrames,
      shouldCapture: shouldCapture,
      corners: mapped(displayed ?? previous))
  }

  func onCaptured(timestamp: UInt64) {
    cooldownUntil = timestamp + cooldownNanoseconds
    capturedCurrentDocument = true
  }

  func reset() {
    resetCandidate()
    cooldownUntil = 0
    capturedCurrentDocument = false
  }

  private func resetCandidate() {
    previous = nil
    displayed = nil
    pendingMovement = nil
    stableSince = 0
    stableFrames = 0
    lastDetectedAt = 0
  }

  private func flattened(_ corners: [[String: NSNumber]]?) -> [Double]? {
    guard let corners, corners.count == 4 else { return nil }
    var values: [Double] = []
    for point in corners {
      guard let x = point["x"]?.doubleValue, let y = point["y"]?.doubleValue else {
        return nil
      }
      values.append(x)
      values.append(y)
    }
    return values
  }

  private func mapped(_ values: [Double]?) -> [[String: NSNumber]]? {
    guard let values, values.count == 8 else { return nil }
    return stride(from: 0, to: values.count, by: 2).map { index in
      [
        "x": NSNumber(value: values[index]),
        "y": NSNumber(value: values[index + 1]),
      ]
    }
  }

  private func smoothed(_ current: [Double]?, toward target: [Double]) -> [Double] {
    guard let current, current.count == target.count else { return target }
    let factor = 0.20
    return zip(current, target).map { source, destination in
      source + (destination - source) * factor
    }
  }

  private func maximumMovement(
    _ first: [Double],
    _ second: [Double],
    _ width: Int,
    _ height: Int
  ) -> Double {
    // Preserve AutoScanHandler's iOS matching rule. It intentionally tracks
    // one inward-facing axis per ordered corner, which ignores contour jitter
    // along the document edges while still resetting for real movement.
    return max(
      abs(first[0] - second[0]) * Double(width),
      abs(first[3] - second[3]) * Double(height),
      abs(first[4] - second[4]) * Double(width),
      abs(first[7] - second[7]) * Double(height)
    )
  }

  private func number(_ options: [String: Any]?, _ key: String, _ fallback: Double) -> Double {
    (options?[key] as? NSNumber)?.doubleValue ?? fallback
  }

  private func milliseconds(_ value: Double, minimum: Double) -> UInt64 {
    UInt64(max(minimum, value) * 1_000_000)
  }
}

File: document_scanner_flutter/android/src/main/kotlin/br/com/dinheironanota/document_scanner_flutter/StabilityTracker.kt
package br.com.dinheironanota.document_scanner_flutter

internal data class StabilityUpdate(
    val state: String,
    val progress: Double,
    val stableFrames: Long,
    val shouldCapture: Boolean,
    val corners: DoubleArray? = null,
)

/**
 * Time-based stability tracker matching the legacy AutoScanHandler contract.
 *
 * A candidate must stay within [distanceThreshold] for [delayMs] and then for
 * [durationMs]. Movement is measured in oriented analysis-image pixels.
 */
internal class StabilityTracker(options: Map<String, Any?>?) {
    private var distanceThreshold = options.number("autoCaptureDistanceThreshold", 50.0)
    private var delayMs = options.number("autoCaptureDelayMs", 1000.0).toLong().coerceAtLeast(0)
    private var durationMs = options.number("autoCaptureDurationMs", 1000.0).toLong().coerceAtLeast(1)
    private var cooldownMs = options.number("autoCaptureCooldownMs", 1500.0).toLong().coerceAtLeast(0)

    private var previous: DoubleArray? = null
    private var displayed: DoubleArray? = null
    private var pendingMovement: DoubleArray? = null
    private var stableSinceNs = 0L
    private var stableFrames = 0L
    private var cooldownUntilNs = 0L
    private var capturedCurrentDocument = false

    @Synchronized
    fun updateOptions(options: Map<String, Any?>?) {
        distanceThreshold = options.number("autoCaptureDistanceThreshold", distanceThreshold)
        delayMs = options.number("autoCaptureDelayMs", delayMs.toDouble()).toLong().coerceAtLeast(0)
        durationMs = options.number("autoCaptureDurationMs", durationMs.toDouble()).toLong().coerceAtLeast(1)
        cooldownMs = options.number("autoCaptureCooldownMs", cooldownMs.toDouble()).toLong().coerceAtLeast(0)
    }

    @Synchronized
    fun update(
        corners: DoubleArray?,
        imageWidth: Int,
        imageHeight: Int,
        timestampNs: Long,
        autoCaptureEnabled: Boolean,
    ): StabilityUpdate {
        if (corners == null || corners.size != 8) {
            val hadDocument = previous != null
            resetCandidate()
            capturedCurrentDocument = false
            return StabilityUpdate(if (hadDocument) "lost" else "searching", 0.0, 0, false)
        }

        val prior = previous
        if (prior == null) {
            previous = corners.copyOf()
            displayed = corners.copyOf()
            stableSinceNs = timestampNs
            stableFrames = 1
            return StabilityUpdate(
                state = "detected",
                progress = 0.0,
                stableFrames = stableFrames,
                shouldCapture = false,
                corners = displayed?.copyOf(),
            )
        }

        if (maximumMovement(prior, corners, imageWidth, imageHeight) >= distanceThreshold) {
            val pending = pendingMovement
            if (
                pending != null &&
                maximumMovement(pending, corners, imageWidth, imageHeight) < distanceThreshold
            ) {
                // Match iOS: two consecutive, mutually close frames represent
                // real movement. A single distant contour is detector noise and
                // must not reset the progress.
                previous = corners.copyOf()
                displayed = smoothed(displayed, corners)
                pendingMovement = null
                stableSinceNs = timestampNs
                stableFrames = 1
                capturedCurrentDocument = false
                return StabilityUpdate(
                    state = "detected",
                    progress = 0.0,
                    stableFrames = stableFrames,
                    shouldCapture = false,
                    corners = displayed?.copyOf(),
                )
            }
            pendingMovement = corners.copyOf()
            return progressUpdate(
                timestampNs = timestampNs,
                autoCaptureEnabled = autoCaptureEnabled,
                allowCapture = false,
            )
        }

        pendingMovement = null
        previous = corners.copyOf()
        displayed = smoothed(displayed, corners)
        stableFrames += 1
        return progressUpdate(
            timestampNs = timestampNs,
            autoCaptureEnabled = autoCaptureEnabled,
            allowCapture = true,
        )
    }

    private fun progressUpdate(
        timestampNs: Long,
        autoCaptureEnabled: Boolean,
        allowCapture: Boolean,
    ): StabilityUpdate {
        val elapsedMs = (timestampNs - stableSinceNs).coerceAtLeast(0) / 1_000_000.0
        if (elapsedMs < delayMs) {
            return StabilityUpdate(
                state = "detected",
                progress = 0.0,
                stableFrames = stableFrames,
                shouldCapture = false,
                corners = displayed?.copyOf() ?: previous?.copyOf(),
            )
        }

        val scanningElapsedMs = elapsedMs - delayMs
        val progress = (scanningElapsedMs / durationMs).coerceIn(0.0, 1.0)
        if (scanningElapsedMs < durationMs) {
            return StabilityUpdate(
                state = "stabilizing",
                progress = progress,
                stableFrames = stableFrames,
                shouldCapture = false,
                corners = displayed?.copyOf() ?: previous?.copyOf(),
            )
        }

        val shouldCapture = allowCapture &&
            autoCaptureEnabled &&
            !capturedCurrentDocument &&
            timestampNs >= cooldownUntilNs
        if (shouldCapture) capturedCurrentDocument = true
        return StabilityUpdate(
            state = "stable",
            progress = 1.0,
            stableFrames = stableFrames,
            shouldCapture = shouldCapture,
            corners = displayed?.copyOf() ?: previous?.copyOf(),
        )
    }

    @Synchronized
    fun onCaptured(timestampNs: Long) {
        cooldownUntilNs = timestampNs + cooldownMs * 1_000_000L
        capturedCurrentDocument = true
    }

    @Synchronized
    fun reset() {
        resetCandidate()
        capturedCurrentDocument = false
        cooldownUntilNs = 0
    }

    private fun resetCandidate() {
        previous = null
        displayed = null
        pendingMovement = null
        stableSinceNs = 0
        stableFrames = 0
    }

    private fun smoothed(current: DoubleArray?, target: DoubleArray): DoubleArray {
        if (current == null || current.size != target.size) return target.copyOf()
        val factor = 0.35
        return DoubleArray(target.size) { index ->
            current[index] + (target[index] - current[index]) * factor
        }
    }

    private fun maximumMovement(
        first: DoubleArray,
        second: DoubleArray,
        width: Int,
        height: Int,
    ): Double =
        // Preserve the legacy iOS matching rule: track one inward-facing
        // axis per ordered corner so edge jitter does not look like movement.
        maxOf(
            kotlin.math.abs(first[0] - second[0]) * width,
            kotlin.math.abs(first[3] - second[3]) * height,
            kotlin.math.abs(first[4] - second[4]) * width,
            kotlin.math.abs(first[7] - second[7]) * height,
        )

    private fun Map<String, Any?>?.number(name: String, fallback: Double): Double =
        (this?.get(name) as? Number)?.toDouble() ?: fallback
}
